libs/langchain/src/agents/middlewareAgent/middleware/bigTool.ts

```diff
@@ -0,0 +1,239 @@
+
+import { createMiddleware } from "../middleware.js";
+import { initChatModel } from "../../../chat_models/universal.js";
+import type { ModelRequest } from "../types.js";
+
+/**
+ * Zod schema for tool selection structured output.
+ */
+const ToolSelectionSchema = z.object({
+  selectedTools: z.array(z.string()).describe("List of selected tool names"),
+});
+
+const DEFAULT_SYSTEM_PROMPT =
+  "Your goal is to select the most relevant tool for answering the user's query.";
+const DEFAULT_INCLUDE_FULL_HISTORY = false;
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * Options for configuring the LLM Tool Selector middleware.
+ */
+export const LLMToolSelectorOptionsSchema = z.object({
+  /**
+   * The language model to use for tool selection (default: the provided model from the agent options).
+   */
+  model: z.string().or(z.instanceof(BaseChatModel)).optional(),
+  /**
+   * System prompt for the tool selection model.
+   */
+  systemPrompt: z.string().default(DEFAULT_SYSTEM_PROMPT),
+  /**
+   * Maximum number of tools to select.
+   */
+  maxTools: z.number().optional(),
+  /**
+   * Whether to include the full conversation history in the tool selection prompt.
+   */
+  includeFullHistory: z.boolean().default(DEFAULT_INCLUDE_FULL_HISTORY),
+  /**
+   * Maximum number of retries if the model selects incorrect tools.
+   */
+  maxRetries: z.number().default(DEFAULT_MAX_RETRIES),
+});
+export type LLMToolSelectorConfig = InferInteropZodInput<
+  typeof LLMToolSelectorOptionsSchema
+>;
+
+/**
+ * Middleware for selecting tools using an LLM-based strategy.
+ *
+ * This middleware analyzes the user's query and available tools to select
+ * the most relevant tools for the task, reducing the cognitive load on the
+ * main model and improving response quality.
+ *
+ * @param options - Configuration options for the middleware
+ * @param options.model - The language model to use for tool selection (default: the provided model from the agent options).
+ * @param options.systemPrompt - System prompt for the tool selection model.
+ * @param options.maxTools - Maximum number of tools to select.
+ * @param options.includeFullHistory - Whether to include the full conversation history in the tool selection prompt.
+ * @param options.maxRetries - Maximum number of retries if the model selects incorrect tools.
+ *
+ * @example
+ * ```ts
+ * import { llmToolSelectorMiddleware } from "langchain/agents/middleware";
+ *
+ * const middleware = llmToolSelectorMiddleware({
+ *   maxTools: 3,
+ *   systemPrompt: "Select the most relevant tools for the user's query."
+ * });
+ *
+ * const agent = createAgent({
+ *   model: "openai:gpt-4o",
+ *   tools: [tool1, tool2, tool3, tool4, tool5],
+ *   middleware: [middleware],
+ * });
+ * ```
+ */
+export function llmToolSelectorMiddleware(options: LLMToolSelectorConfig) {
+  return createMiddleware({
+    name: "LLMToolSelector",
+    contextSchema: LLMToolSelectorOptionsSchema,
```

Comment:
We're missing here an option to always include some tools.

A user may want to make sure that some tools are always included (and these tools should be removed from the prompt; i.e., the model shouldn't necessarily know about them.)